crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -4096,6 +4193,93 @@ mod tests {
         Ok(())
     }
 
+    #[test]
+    fn except_handler() -> anyhow::Result<()> {
+        let mut db = setup_db();
+
+        db.write_dedented(
+            "src/a.py",
+            "
+            import builtins
+
+            try:
+                x
+            except NameError as e:
+                pass
+            except (TypeError, builtins.AttributeError) as f:
+                pass
+            ",
+        )?;
+
+        assert_public_ty(&db, "src/a.py", "e", "NameError");
+        assert_public_ty(&db, "src/a.py", "f", "TypeError | AttributeError");
+        assert_file_diagnostics(&db, "src/a.py", &[]);
+
+        Ok(())
+    }
+
+    #[test]
+    fn invalid_except_handler() -> anyhow::Result<()> {
+        let mut db = setup_db();
+
+        db.write_dedented(
+            "src/a.py",
+            "
+            try:
+                x
+            except NameError() as e:
+                pass
+            except (TypeError, AttributeError()) as f:
+                pass
+            ",
+        )?;
+
+        assert_file_diagnostics(
+            &db,
+            "src/a.py",
+            &[
+                "Can only catch a `BaseException` subclass or a tuple of `BaseException` subclasses, not an object of type 'NameError'",
+                "Can only catch a `BaseException` subclass or a tuple of `BaseException` subclasses, not an object of type 'AttributeError'"
+            ]
+        );
+        assert_public_ty(&db, "src/a.py", "e", "Unknown");
```

Comment:
`foo` isn't necessarily an invalid object to be caught by an exception handler, though. We just don't know, because we weren't able to resolve `nonexistent_module` and inspect where it came from. It could be perfectly valid (maybe `nonexistent_module` is a C extension that doesn't have stubs), and we've already emitted an error at the import statement complaining that we couldn't resolve the import.

Inferring `Never` for `e` could create lots of false-positive errors when the `e` variable is actually used, since `Never` has no attributes or methods.

Reply:
FWIW, both mypy and pyright infer `Any`/`Unknown` in this situation:

- https://mypy-play.net/?mypy=latest&python=3.12&gist=bcdb068154c1fa6bac71335781c392b4
- https://pyright-play.net/?pythonVersion=3.13&strict=true&code=GYJw9gtgBAJmCmBnA%2BgOzAF2fAHgS0QyjwgAcwQjgwwBYAKAYxAE8AuBqLqHB3AY3ikqNKAENEUeB3rcoIeADd4YgDbIMLUvAAU8AJQMgA

Reply:
Oh, I think I just commented on the wrong case! `foo` should be `Unknown` here, and I definitely agree that if one of the caught exception types is `Unknown`, that should be valid and should contribute `Unknown` to the resulting union.

Where I think we might want `Never` is for types that we _know_ are not `BaseException` subtypes; the cases where the previous version of this PR emitted a diagnostic. But this version of the PR no longer attempts to detect such cases at all (which I think makes sense), so it's a moot question for this PR.

Reply:
Gotcha! I think we're in agreement. Thanks for clarifying!